### DISCUSSION TITLE
How to properly dynamically create shapes like arcs?

### DISCUSSION THREAD

**Discussion by stepancheg** [bevyengine/bevy#10923] [Q&A]
I need to draw many (millions) of 3D arcs of arbitrary sizes (10º, 23º, 57.456º etc), including dynamically. But not all of them are on the screen at the same time.

Something like this:

<img width="383" alt="image" src="https://github.com/bevyengine/bevy/assets/28969/dc3bed4e-755b-4e49-991d-7a067418afd7">

(They are arcs now, perhaps will be bezier curves later, I don't know)

`Transform` can only do affine transformations. So I cannot create for example a basic arc and skew it it into an arc I need.

So, after hours reading Bevy code and examples, I see these options:
* Generate 1000 arcs for each length with 0.1º step and hope artifacts won't be visible
* Generate new mesh resources each time I need to draw more arcs
* Use skeletal animation framework to bend an arc a little. No actual animation
* Write a vertex shader

So
* option 1 (1000 arcs), need experimenting to see how bad it would be with large zoom for example
* option 2 (new mesh) is the easiest, but I don't understand how bad it would be for performance. I'm thinking: it is stored as resource, not in entity, so it is not meant to be created dynamically?
* option 3 (skeletal animation) seems to be good from performance point of view, but requires understanding how skeletal animation works (still trying to understand `SkinnedMeshInverseBindposes`)
* option 4 (vertex shader) seems to be perfect for performance, and logic is easy, but I don't even know where to start reading. Pointers?

I don't expect detailed answers, but some pointers or just intuition would be appreciated.

Thanks!

**Comment by nicopap** [comment]
You have a few options here:

* Bevy 0.12 has a new ["standard material extension" feature](https://bevyengine.org/news/bevy-0-12/#material-extensions). You could create your own material for arcs. Disclaimer: I've not used myself, and I'm not 100% it is possible to use for this usecase.
* Fork and update [bevy_polyline](https://github.com/ForesightMiningSoftwareCorporation/bevy_polyline/), though you would need to add handling for linecaps, otherwise you wouldn't get nice arcs.
* Use [bevy_vector_shape](https://github.com/james-j-obrien/bevy_vector_shapes). Never used it and not exactly sure it's pertinent to this use case either.

Honestly it all depends on your use case. Do you mean to draw and combine "bendable" lines? Is the shape of the line to be updated each frame, or the same line will persist, but their `Transform` change? Are those in a 3D context or in 2D? etc. I can't recommend a specific approach without more context.

**Reply by stepancheg** [reply]
So, what I got from reading these three links.

Material can override "vertex shader", but not "geometry shader". So it is quite low level. And the other two seems to use vertex shaders.

> Do you mean to draw and combine "bendable" lines?

Probably not.

> Is the shape of the line to be updated each frame, or the same line will persist, but their Transform change?

No, they are mostly static. On some frame I can have a thousand added, on another frame a thousand removed, but in average frame nothing changes.

> Are those in a 3D context or in 2D?

3D.

> I can't recommend a specific approach without more context.

No problem. Thank you, I learned a few new things today, and got a lot of code to read.

**Reply by nicopap** [reply]
yeah sounds like you should use `bevy_polyline`. It sounds like it fits perfectly your usecase. Though you'll have to update it to work with bevy 0.12, add joint handling and frustum culling.

**Comment by rlidwka** [comment]
I guess you can't use gizmos? That would be the most obvious solution.

> option 2 (new mesh) is the easiest, but I don't understand how bad it would be for performance.

Try it and find out. As far as I remember, you can also edit meshes at runtime, so you can try to reuse existing ones if it helps.

**Reply by stepancheg** [reply]
> I guess you can't use gizmos?

I guess gizmos are expensive. No optimizations like do not do anything if the object is out of sight.

> Try it and find out.

I was hoping to save some time by choosing a path which is more likely to be successful. But this is always a backup option.
